jobs,server: graceful shutdown for secondary tenant servers

knz opened the pull request:
Epic: CRDB-23559
Fixes #92523.

All commits but the last are from #100436.

This change ensures that tenant servers managed by the server
controller receive a graceful drain request as part of the graceful
drain process of the surrounding KV node.

This change, in turn, ensures that SQL clients connected to these
secondary tenant servers benefit from the same guarantees (and
graceful periods) as clients to the system tenant.

knz committed:
server: gracefully shut down secondary tenant servers

This change ensures that tenant servers managed by the server
controller receive a graceful drain request as part of the graceful
drain process of the surrounding KV node.

This change, in turn, ensures that SQL clients connected to these
secondary tenant servers benefit from the same guarantees (and
graceful periods) as clients to the system tenant.

Release note: None

stevendanna commented on pkg/jobs/registry.go:
```diff
@@ -1066,6 +1071,9 @@ func (r *Registry) Start(ctx context.Context, stopper *stop.Stopper) error {
 	})
 
 	if err := stopper.RunAsyncTask(ctx, "jobs/cancel", func(ctx context.Context) {
+		r.startedControllerTasksWG.Add(1)
```
We have examples of both patterns in the database, but I almost always put the Add() before launching the goroutine/async task.  You don't have a guarantee that the scheduler has started the goroutine before you end up calling Wait().

stevendanna commented on pkg/server/server_controller_orchestration.go:
```diff
@@ -483,3 +538,47 @@ func (c *serverController) logStopEvent(
+
+	drainCtx := logtags.WithTags(context.Background(), logtags.FromContext(ctx))
+
+	for ; ; prevRemaining = remaining {
```
[nit] I have a feeling this would become a little clearer to read if remaining became local to the scope of the block, and we set `prevRemaining = remaining` at the end of the loop, and then used `for {` here.

stevendanna commented on pkg/server/server_controller_orchestration.go:
```diff
@@ -212,6 +213,28 @@ func (c *serverController) startControlledServer(
+		}
+	}
+	gracefulDrain := func(ctx context.Context) {
+		drainServer := func() interface {
+			gracefulDrain(ctx context.Context, verbose bool) (uint64, redact.RedactableString, error)
+		} {
+			shutdownInterface.Lock()
+			defer shutdownInterface.Unlock()
+			return shutdownInterface.drainableServer
+		}()
```
[nit] Perhaps I'm missing something, but this is a lot of readability overhead. I know we like defers for locks, but in this case

```
shutdownInterface.Lock()
drainServer := shutdownInterface.drainableServer
shutdownInterface.Unlock()
````

feels pretty attractive. 

Overall, I think we would benefit from refactoring this a bit so that we had some set of methods on serverEntry or some new type that wrapped some of these statement management things up in some functions so they didn't have to all be inline to this function.

stevendanna reviewed (commented):
Overall, this looks reasonable to me.  `startControlledServer` is a bit of a beast, but that is a fish to fry for another day.

knz reviewed (commented):
>   startControlledServer is a bit of a beast, but that is a fish to fry for another day.

I reduced the quantity of changes to the function. PTAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99958#-:-NRrnHAAEuqaN02xAIvk:b-xxge0m)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @JeffSwenson and @stevendanna)

___
*[`pkg/jobs/registry.go` line 1074 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/99958#-NRiVSoNIIgD2P2D1uk6-r2-1074:-NRni-K17l-PUOm6hwTi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/75a863ce5f572d9e21dc8021693bcdc9fc819bf4/pkg/jobs/registry.go#L1074)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

We have examples of both patterns in the database, but I almost always put the Add() before launching the goroutine/async task.  You don't have a guarantee that the scheduler has started the goroutine before you end up calling Wait().
</blockquote></details>

Done.
___
*[`pkg/server/server_controller_orchestration.go` line 229 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/99958#-NRiVSoOiqlsMJStryqt-r2-229:-NRrnFMAA2PNOF5nSleS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/75a863ce5f572d9e21dc8021693bcdc9fc819bf4/pkg/server/server_controller_orchestration.go#L229)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

As a quick follow up, the refactor I mentioned is strictly future looking. Not required for this PR.
</blockquote></details>

Done.
___
*[`pkg/server/server_controller_orchestration.go` line 561 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/99958#-NRiVSoOiqlsMJStryqt-r2-561:-NRrhl30-B1k0cIuMNym:bkvylip) ([raw file](https://github.com/cockroachdb/cockroach/blob/75a863ce5f572d9e21dc8021693bcdc9fc819bf4/pkg/server/server_controller_orchestration.go#L561)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

[nit] I have a feeling this would become a little clearer to read if remaining became local to the scope of the block, and we set `prevRemaining = remaining` at the end of the loop, and then used `for {` here.
</blockquote></details>

Done.

Also I made this code shared with pkg/cli/start.go as suggested earlier.


<!-- Sent from Reviewable.io -->

stevendanna commented on pkg/server/server_controller_orchestration.go:
```diff
@@ -204,39 +212,64 @@ func (c *serverController) startControlledServer(
+			// by the tenant server any more.
+			drainCtx, cancel := c.stopper.WithCancelOnQuiesce(tenantCtx)
+			defer cancel()
+			shutdownInterface.maybeCallDrain(drainCtx)
```
I think we probably have other problems of this nature so perhaps it isn't worth fixing, but I think the current structure will fail to drain the server if the following sequence happens:

1) The propogate-close task starts and is running
2) The managed-tenant-server task starts and reaches line 349 but has not run line 350
3) createServerEntryLocked returns the server entry, the entry is added to the server slice, the severs slice mutex is unlocked.
4) requestAllStopped() is called.
5) propogate-close sees the stoprequest, and runs myabeCallDrain(), but we haven't set the server yet, so it reports nothing to drain, even though our server already has drainable components started.

I think this is minor because the server won't be serving any user connections yet (since the orchestrator doesn't think it is started). So it just means any jobs or other internal components might not get drained.

stevendanna commented on pkg/jobs/registry.go:
```diff
@@ -1915,14 +1945,25 @@ func (r *Registry) IsDraining() bool {
+
 // SetDraining informs the job system if the node is draining.
 //
 // NB: Check the implementation of drain before adding code that would
 // make this block.
```
[nit[ We can probably remove this comment now

stevendanna reviewed (commented):
Left a few nits and one minor race/leak.  I'll leave the latter to your judgement.

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99958)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @JeffSwenson, @mgartner, and @stevendanna)

___
*[`pkg/jobs/registry.go` line 1958 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/99958#-NRiVSoNIIgD2P2D1uk6-r6-1958:-NRvqPvp8Q-niR-VVddp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/83edc7f0c33013b1a963c1c05a72a58b0d4410ae/pkg/jobs/registry.go#L1958)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

[nit[ We can probably remove this comment now 
</blockquote></details>

Done.
___
*[`pkg/server/drain.go` line 405 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/99958#-NRiVSoOiqlsMJStryqp-r6-405:-NRvr2E-AreXhU3vIPlE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/83edc7f0c33013b1a963c1c05a72a58b0d4410ae/pkg/server/drain.go#L405)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

[nit] ode is draining
</blockquote></details>

Done.
___
*[`pkg/server/server_controller_orchestration.go` line 257 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/99958#-NRiVSoOiqlsMJStryqt-r6-257:-NRvrZO54QWOfeIcV-ns:b-8raz56) ([raw file](https://github.com/cockroachdb/cockroach/blob/83edc7f0c33013b1a963c1c05a72a58b0d4410ae/pkg/server/server_controller_orchestration.go#L257)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

I think we probably have other problems of this nature so perhaps it isn't worth fixing, but I think the current structure will fail to drain the server if the following sequence happens:

1) The propogate-close task starts and is running
2) The managed-tenant-server task starts and reaches line 349 but has not run line 350
3) createServerEntryLocked returns the server entry, the entry is added to the server slice, the severs slice mutex is unlocked.
4) requestAllStopped() is called.
5) propogate-close sees the stoprequest, and runs myabeCallDrain(), but we haven't set the server yet, so it reports nothing to drain, even though our server already has drainable components started.

I think this is minor because the server won't be serving any user connections yet (since the orchestrator doesn't think it is started). So it just means any jobs or other internal components might not get drained.


</blockquote></details>

I'm impressed and curious how you went about analyzing this condition :100: 

The solution here is to split the instantiation of a new server and the starting of the new server in two separate steps (i.e. split `(s *Server) startTenantServerInternal`).  We can then have the shutdown interface ready after instantiation and before start. This incidentally mirrors the dance done in cli/start.go for exactly the same reason.

I believe this is needed anyway to solve  #97661 /  #98868. I'll mull it over.
___
*[`pkg/sql/stats/automatic_stats.go` line 378 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/99958#-NRsG3Bl6JtUsYDhCDAL-r6-378:-NRvr3Pi45F9zaj8lDh8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/83edc7f0c33013b1a963c1c05a72a58b0d4410ae/pkg/sql/stats/automatic_stats.go#L378)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Stray comment
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

knz commented:
TODO for self: add a warning in logs when a tenant server is shut down without graceful drain.

knz committed:
server: unexport some functions

Release note: None

knz committed:
server: save a reference to BaseConfig in SQLServerWrapper

Release note: None

knz committed:
server: extend the `onDemandServer` interface

ahead of splitting "new" vs "start" during construction.

Release note: None

knz committed:
server: lift `reportTenantInfo` into `*SQLServerWrapper`

Release note: None

knz committed:
server: separate "new" from "start" for tenant servers

This peels the call to "start" from the `newTenantServer`
interface and pulls it into the orchestration retry loop.

This change also incidentally reveals an earlier misdesign: we are
calling `newTenantServer` _then_ `start` in the same retry loop. If
`new` succeeds but `start` fails, the next retry will call
`newTenantServer` again *with the same stopper*, which will leak
closers from the previous call to `new`.

Release note: None

knz committed:
server: fix the tenant server error handling

Prior to this patch, if an error occurred during the initialization or
startup of a secondary tenant server, the initialization would leak
state into the stopper defined for that tenant. Generally, reusing
a stopper across server startup failures is not safe (and API
violation).

This patch fixes it by decoupling the intermediate stopper used for
orchestration from the one used per tenant server.

Release note: None

knz committed:
serverccl: simplify TestServerStartupGuardrails

Prior to this patch, the test was not cleaning up its server stopper
reliably at the end of each sub-test. This patch fixes it.

Release note: None

knz reviewed (commented):
Rebased on top of #100436.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99958#-:-NS2P3Rg9HnpHh4bzXrJ:b-cr8m6x)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @JeffSwenson, @mgartner, and @stevendanna)

___
*[`pkg/server/server_controller_orchestration.go` line 257 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/99958#-NRiVSoOiqlsMJStryqt-r6-257:-NS2Ozki179rCzABOlg0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/83edc7f0c33013b1a963c1c05a72a58b0d4410ae/pkg/server/server_controller_orchestration.go#L257)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

I'm impressed and curious how you went about analyzing this condition :100: 

The solution here is to split the instantiation of a new server and the starting of the new server in two separate steps (i.e. split `(s *Server) startTenantServerInternal`).  We can then have the shutdown interface ready after instantiation and before start. This incidentally mirrors the dance done in cli/start.go for exactly the same reason.

I believe this is needed anyway to solve  #97661 /  #98868. I'll mull it over.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

knz committed:
serverccl: make TestServerControllerMultiNodeTenantStartup faster

Release note: None

knz committed:
serverccl: clarify the progress inside TestServerControllerMultiNodeTenantStartup

Release note: None

knz commented:
note to self: add a unit test using the new test knob

knz commented:
Note to self: abandon drain on stopper quiescing

knz commented:
bors r=stevendanna
